packages/block-editor/src/store/reducer.js

```diff
@@ -0,0 +1,893 @@
+ * @param {Function} transform Transforming function to be applied to each block.
+ *
+ * @return {Object} Flattened object.
+ */
+function flattenBlocks( blocks, transform ) {
+	const result = {};
+
+	const stack = [ ...blocks ];
+	while ( stack.length ) {
+		const { innerBlocks, ...block } = stack.shift();
+		stack.push( ...innerBlocks );
+		result[ block.clientId ] = transform( block );
+	}
+
+	return result;
+}
+
+/**
+ * Given an array of blocks, returns an object containing all blocks, without
+ * attributes, recursing into inner blocks. Keys correspond to the block client
+ * ID, the value of which is the attributes object.
+ *
+ * @param {Array} blocks Blocks to flatten.
+ *
+ * @return {Object} Flattened block attributes object.
+ */
+function getFlattenedBlocksWithoutAttributes( blocks ) {
+	return flattenBlocks( blocks, ( block ) => omit( block, 'attributes' ) );
+}
+
+/**
+ * Given an array of blocks, returns an object containing all block attributes,
+ * recursing into inner blocks. Keys correspond to the block client ID, the
+ * value of which is the attributes object.
+ *
+ * @param {Array} blocks Blocks to flatten.
+ *
+ * @return {Object} Flattened block attributes object.
+ */
+function getFlattenedBlockAttributes( blocks ) {
+	return flattenBlocks( blocks, ( block ) => block.attributes );
+}
+
+/**
+ * Given a block order map object, returns *all* of the block client IDs that are
+ * a descendant of the given root client ID.
+ *
+ * Calling this with `rootClientId` set to `''` results in a list of client IDs
+ * that are in the post. That is, it excludes blocks like fetched reusable
+ * blocks which are stored into state but not visible.
+ *
+ * @param {Object}  blocksOrder  Object that maps block client IDs to a list of
+ *                               nested block client IDs.
+ * @param {?string} rootClientId The root client ID to search. Defaults to ''.
+ *
+ * @return {Array} List of descendant client IDs.
+ */
+function getNestedBlockClientIds( blocksOrder, rootClientId = '' ) {
+	return reduce( blocksOrder[ rootClientId ], ( result, clientId ) => [
+		...result,
+		clientId,
+		...getNestedBlockClientIds( blocksOrder, clientId ),
+	], [] );
+}
+
+/**
+ * Returns an object against which it is safe to perform mutating operations,
+ * given the original object and its current working copy.
+ *
+ * @param {Object} original Original object.
+ * @param {Object} working  Working object.
+ *
+ * @return {Object} Mutation-safe object.
+ */
+function getMutateSafeObject( original, working ) {
+	if ( original === working ) {
+		return { ...original };
+	}
+
+	return working;
+}
+
+/**
+ * Returns true if the two object arguments have the same keys, or false
+ * otherwise.
+ *
+ * @param {Object} a First object.
+ * @param {Object} b Second object.
+ *
+ * @return {boolean} Whether the two objects have the same keys.
+ */
+export function hasSameKeys( a, b ) {
+	return isEqual( Object.keys( a ), Object.keys( b ) );
```

Comment:
Don't we need to account for `undefined` attributes here?

Reply:
> Don't we need to account for `undefined` attributes here?

Yes and no.

Yes in that, if passed `undefined` objects, this will throw an uncaught error. This was ultimately behind some previous E2E test failures, and to that point it's not worth risking regressions by changing it from how it [behaves in the master branch](https://github.com/WordPress/gutenberg/blob/011a8b3d92f2bdd8b79853bb53d292999dc4c5dd/packages/editor/src/store/reducer.js#L168-L179). It also exists in `editor/reducer.js` still, where it's currently out of alignment.

No in that the true cause of the failures was tracked back to a different bug, reflected in the change of d904406 . I was curious why it'd ever be the case that `updateBlockAttributes` would be called with an `undefined` attributes object. It turns out that we weren't always comparing two `updateBlockAttributes` action objects, but rather any two actions which had a matching `clientId` and where at least the second action was from `updateBlockAttributes`. The new unit tests should show quite clearly what was happening.

In terms of action items, I'd suggest for the purpose of moving forward:

1. Keep d904406325829a57c7cb20f8568fdb5706f65f9b , though I'd be open to extracting it as a separate pull request fix, since the same problem exists in `master` technically
2. Restore this line of code to how we have it in `editor/reducer.js`, using `keys` in place of `Object.keys`.

In the future, I think we should open the discussion to how we handle cases where action properties passed as arguments by developers are to be tolerated. We're not very consistent with this, and it's pretty easy to produce errors when calling actions unexpectedly:

```js
var clientId = wp.data.select( 'core/editor' ).getBlockOrder()[ 0 ]
wp.data.dispatch( 'core/editor' ).updateBlock( clientId )
// Uncaught (in promise) TypeError: Cannot read property 'attributes' of undefined
```

We should consider:

- Always treating them as "maybe undefined" in reducer logic, consistently
- Defaulting any action property value early in the reducer implementation
- Defaulting any action argument in the action function itself

Reply:
>1\. Keep [d904406](https://github.com/WordPress/gutenberg/commit/d904406325829a57c7cb20f8568fdb5706f65f9b) , though I'd be open to extracting it as a separate pull request fix, since the same problem exists in `master` technically

It's actually not an issue in `master`, because the condition occurs earlier in the logic before reaching `isUpdatingSameBlockAttribute`.

https://github.com/WordPress/gutenberg/blob/211f65dd854fc1f7d7e54782863ad0629abeff41/packages/editor/src/store/reducer.js#L226-L235

To that point, then, we'll need a similar condition somewhere to check that `lastAction` _exists_.

>2\. Restore this line of code to how we have it in `editor/reducer.js`, using `keys` in place of `Object.keys`.

Aligned in 724b38f